File: hw1/kmeans.c
/*
 * Kmeans implementation in C
 * Software Project ex. 1
 *
 * Amit Elyasi
 * 316291434
 *
 * Nizan Shemi
 * 206962912
 *
 */

#include <stdio.h>
#include <stdlib.h>


/*
 * implementations of kmeans algorithm:
 * given n data point (in R^d), group the data into k clusters,
 * each data point is assigned to exacly one cluster (note that k<n)
 */
int kmeans(int k, float* data_points, float* centroids, float* utl ,int max_iter, int dim, int n){
    short convergece = 1;
    void assign(float*, float*, int, int, int);
    short re_estimate(float*, float* , float*, int, int, int);
    void build_clusters(int, int, float*, float*);
    int i;

    build_clusters(k, dim, data_points, centroids);
    for (i=0; i<max_iter; i++){
        assign(data_points, centroids, dim, n, k);
        convergece = re_estimate(data_points, centroids, utl, dim, n, k);
        if (convergece == 1) {
            return 0;
        }
    }
    return 0;
}


/*
 * assigns data points to their closest cluster (measure distance from the centroid)
 * updates the number of cluster for each data point
 */
void assign(float* data_points, float* clusters, int dim, int n, int k){
    int INT_MAX = 2147483647;
    int cluster;
    float distance(float *, float *, int, int, int);
    int v,c;
    float min_dis, dis;
    void set(float *, int, int , int,float);

    min_dis = INT_MAX;
    for(v = 0; v < n; v++){
        for(c = 0;c < k; c++){
            dis = distance(data_points, clusters, dim, v, c);
            if( dis <= min_dis){
                min_dis = dis;
                cluster = c;
            }
        }
        set(data_points, v, dim, dim + 1, cluster);
        min_dis = INT_MAX;
    }
}


/*
 * re-estimates a centroid for each cluster:
 * for each cluster calculate the average of the points assign to it,
 * updates centroids to be the average vector,
 * returns 1 if the old centroids are equal to the new ones.
 */
short re_estimate(float* data_points, float* clusters,float *utl, int dim, int n, int k) {
    void vec_sum(float* , float* , int, int, int);
    void zero_mat(float *, int, int);
    float get(float *, int, int, int);
    void set(float *, int, int, int, float);
    short isEqual = 1;
    int i, j;
    float x;

    zero_mat(utl, dim + 1, k);

    /* sum all vectors for each cluster */
    for (i = 0; i < n; i++) {
        j = get(data_points, i, dim, dim+1);
        vec_sum(utl, data_points, dim, j, i);
        x = get(utl, j, dim, dim + 1) + 1;
        set(utl, j, dim, dim + 1, x);
    }

    /* Divides each sum by the number of vectors to get average */
    for (i = 0; i < k; i++) {
        for (j = 0; j < dim; j++) {
            x = get(utl, i, j, dim+1);
            set(utl, i, j, dim + 1, (x / get(utl, i, dim, dim+1)));
        }
    }

    /* Compare the old centroids to the new ones */
    for (i = 0; i < k; i++) {
        for (j = 0; j < dim; j++) {
            if (!(((get(clusters, i, j, dim) - get(utl, i, j, dim+1)) < 0.000001) && ((get(clusters, i, j, dim) - get(utl, i, j, dim+1)) > -0.000001))) {
                isEqual = 0;
                break;
            }
        }
        if (!isEqual){
            break;
        }
    }

    /* if there is no change in centroid- we have reach convergence */
    if (isEqual == 1) {
        return 1;
    }

    /* copy the new centroids to the old ones place */
    for (i = 0; i < k; i++) {
        for (j = 0; j < dim; j++) {
            x = get(utl, i, j, dim+1);
            set(clusters, i, j, dim, x);
        }
    }
    return isEqual;
}


/*
* calculate distance between two vectors (v1-v2)^2
*/
float distance(float *v1, float *v2, int dim,int row_v1, int row_v2){
    int i;
    float result = 0;
    float x;
    float get(float *, int, int, int);

    for(i = 0;i < dim;i++){
        x = (get(v1, row_v1, i, dim + 1)-get(v2, row_v2, i, dim));
        x *= x;
        result += x;
    }
    return result;
}


/*
 * adds vec2 to vec1 coordinate wise
 */
void vec_sum(float* vec1, float* vec2, int dim, int row_vec1, int row_vec2){
    float get(float *, int, int, int);
    void set(float *, int ,int, int, float);
    int i;
    float sum;

    for(i = 0;i < dim;i++){
        sum = get(vec1, row_vec1, i, dim+1) + get(vec2, row_vec2, i, dim+1);
        set(vec1, row_vec1, i, dim + 1, sum);
    }
}


/*
 * zeros a given matrix from row start to row end
 */
void zero_mat(float* clusters , int dim, int n){
    int i,j;
    void set(float *, int, int, int, float);

    for(i = 0; i < n; i++){
        for(j=0; j < dim; j++){
            set(clusters, i, j, dim, 0);
        }
    }
}


/* fill given matrix with the data from input file */
void read_data(FILE* fp, float *data_points, char *line, int n, int dim){
    void set(float *, int, int, int, float);
    char *p;
    int size, i, j;

    size = 50 * dim;
    for(i = 0; i < n; i++){

        /* p is a pointer to a beging of a line in the file */
        p = fgets(line, size, fp);

        for(j = 0; j < dim; j++){

            /* extract float form the line */
            set(data_points, i, j, dim + 1, strtod(p, &p));

            p += 1; /* skip comma */
        }
        set(data_points, i, j, dim + 1, 0);
    }
}


/*
 * counts the lines in input file
 */
int num_of_lines(FILE *fp){
    int ch;
    int lines = 0;

    while(!feof(fp))
    {
        ch = fgetc(fp);
        if(ch == '\n')
        {
            lines++;
        }
    }
    return lines;
}


/* count the columns in input file */
int num_of_columns(FILE *fp){
    int ch;
    int columns = 1;

    while(!feof(fp))
    {
        ch = fgetc(fp);
        if(ch == ','){
            columns++;
        }
        if(ch == '\n'){
            break;
        }
    }
    return columns;
}


void build_clusters(int k, int dim, float *vectors, float *centroids){
    void set(float *, int, int, int, float);
    float get(float *, int, int ,int);
    int i, j;

    for(i = 0; i < k; i++){
        for(j = 0;j < dim;j++){
            set(centroids, i, j, dim, get(vectors, i, j, dim + 1));
        }
    }
}


/*
 * prints the centroids in the template requierd
 */
void print_centroids(float* clusters, int k, int dim){
    int i,j;
    float get(float *, int, int, int);

    for(i = 0; i < k;i++){
        for(j = 0;j < dim-1;j++){
            printf("%0.4f,", get(clusters, i, j, dim));
        }
        printf("%.4f\n", get(clusters, i, dim-1, dim));
    }
}


float get(float* arr, int i, int j, int dim){
    int index;

    index = (i*dim + j);
    return arr[index];
}


void set(float* arr, int i, int j, int dim, float item){
    int index;

    index = (i*dim + j);
    arr[index] = item;
}


int main( int argc, char* argv[]) {
    int kmeans(int, float* , float*, float*, int, int, int);
    void read_data(FILE*, float *, char *, int, int );
    void print_centroids(float*, int, int);
    int max_iter, dim, k, n;
    long bOfFile;
    float * utl;
    float *data_points;
    float *centroids;
    char *line;

    /* reading arguments */
    k = strtol(argv[1], NULL, 10);
    if(argc == 3){
        max_iter = strtol(argv[2], NULL, 10);

        if (max_iter<=0){
            printf("INPUT ERROR:\nmaximum iterations is invalid");
            return 1;
        }
    }

    else{
        max_iter = 200;
    }
    bOfFile = ftell(stdin);/*save the address of the beginning of the file */
    n = num_of_lines(stdin);

    if(k<=0){
        printf("INPUT ERROR:\nk is invalid");
        return 1;
    }
    if(n < k){
        printf("INPUT ERROR:\nthere are less then k=%d data points",k);
        return 1;
    }

    fseek(stdin, bOfFile, SEEK_SET);/*set the file position back to the beginning */
    dim = num_of_columns(stdin);
    fseek(stdin, bOfFile, SEEK_SET);/*set the file position back to the beginning */

    line = malloc(sizeof(char) * (30*dim));
    /* build matrix that contins all the points */;
    data_points = malloc(sizeof(float) * ((dim+1) * n));
    /* build matrix that contins all the centroids */;
    centroids = malloc(sizeof(float) * (dim*k));
    /* build matrix that use for calclations */;
    utl = malloc(sizeof(float) * ((dim+1)*k));
    read_data(stdin, data_points, line, n, dim);
    kmeans(k, data_points, centroids, utl, max_iter, dim, n);
    print_centroids(centroids, k, dim);


    /* free the memory used */
    free(line);
    free(data_points);
    free(centroids);
    free(utl);

    return 0;
}

File: hw1/kmeans.py
"""
Kmeans implementation in Python
Software Project ex. 1

Amit Elyasi
316291434

Nizan Shemi
206962912
"""

import sys


def Kmeans(k, max_iter=200):

    data_points = read_data()
    if not good_input(data_points, k):
        return

    centroids = data_points[:k]
    for i in range(max_iter):
        clusters = assign(data_points, centroids)
        new_centroids = re_estimate(clusters)
        if sorted(new_centroids) == sorted(centroids):
            break
        centroids = new_centroids
    return centroids


def assign(data_points, centroids):
    clusters = [[centroid] for centroid in centroids]

    for point in data_points:
        min_dst = distance(point, centroids[0])
        closest_centroid_num = 0

        for i,centroid in enumerate(centroids):
            dst = distance(point, centroid)
            if dst < min_dst:
                min_dst = dst
                closest_centroid_num = i

        clusters[closest_centroid_num].append(point)

    return clusters


def re_estimate(clusters):
    new_centroids = []
    for cluster in clusters:
        cluster = cluster[0:]
        new_centroids.append(average(cluster))

    return new_centroids


def read_data():
    vectors = []
    while(True):
        try:
            for line in input().split("\n"):
                vectors.append(toFloat(line.split(",")))
        except EOFError:
            break
    return vectors


def toFloat(arr):
    for i,str in enumerate(arr):
        arr[i] = float(str)
    return arr


def good_input(vectors, k):
    dim = len(vectors[0])
    for vector in vectors:
        if len(vector) != dim:
            print(f"INPUT ERROR:\nthe vector {vector} is {len(vector)} dimentional, different from the first vector dimention: {dim}")
            return False
    if len(vectors) < k:
        print(f"INPUT ERROR:\nthere are less then k={k} data points")
        return False
    return True


def average(cluster):
    dim = len(cluster[0])
    avg = [0 for i in range(dim)]
    cluster_size = len(cluster)
    for point in cluster:
        avg = vectors_sum(point, avg)
    avg = scalar_div(avg, cluster_size)
    return avg


def distance(vec1, vec2):
    # if vec is scalar:
    if type(vec2) in (int, float):
        return abs(vec1-vec2)

    if len(vec1) != len(vec2):
        print("The vectors are not in the same length")
        return -1
    sum = 0
    for i in range(len(vec1)):
        sum += (vec1[i]-vec2[i])**2
    return sum


def vectors_sum(vec1,vec2):
    # if vec is scalar:
    if type(vec1) in (int, float):
        return vec1+vec2

    if len(vec1) != len(vec2):
        print("The vectors are not in the same length")
        return -1

    sum_vec = [0 for i in range(len(vec1))]
    for i in range(len(vec1)):
        sum_vec[i] = vec1[i] + vec2[i]
    return sum_vec


def scalar_div(vec, x):
    if x==0 :
        print("you can't divide by zero")
        return -1

    # if vec is scalar:
    if type(vec) in (int, float) :
        return vec/x

    for i in range(len(vec)):
        vec[i] = vec[i]/x
    return vec


def main():

    try:
        k = int(sys.argv[1])    
        if (k <= 0):
            print("INPUT ERROR:\nk can't be <= 0")
            return 1
    except ValueError:
        print("INPUT ERROR:\nk can't be a letter")
        return 1

    # algorithm
    if len(sys.argv) >= 3:
        try:
            max_iter = int(sys.argv[2])    

        except ValueError:
            print("INPUT ERROR:\nk can't be a letter")
            return 1
    
        if max_iter <= 0:
            print("INPUT ERROR:\nmax iteration can't be <= 0")
            return 1

        centroids = Kmeans(k, max_iter=max_iter)
    else:
        centroids = Kmeans(k)

    # print in the right format
    for j,centroid in enumerate(centroids):
        if j != 0:
            print("\n", end="")
        for i,coordinate in enumerate(centroid):
            if i!=0 :
                print(",", end="")
            print("%.4f" % coordinate, end="")


if __name__ == "__main__":
    main()
